Source/Core/Core/IOS/FS/FileSystemProxy.cpp

```diff
@@ -0,0 +1,592 @@
+
+#include <algorithm>
+#include <cstring>
+
+#include "Common/Assert.h"
+#include "Common/ChunkFile.h"
+#include "Common/StringUtil.h"
+#include "Common/Swap.h"
+#include "Core/HW/Memmap.h"
+#include "Core/HW/SystemTimers.h"
+#include "Core/IOS/FS/FileSystem.h"
+
+namespace IOS
+{
+namespace HLE
+{
+namespace Device
+{
+using namespace IOS::HLE::FS;
+
+static s32 ConvertResult(ResultCode code)
+{
+  if (code == ResultCode::Success)
+    return IPC_SUCCESS;
+  return -(static_cast<s32>(code) + 100);
+}
+
+static IPCCommandResult GetFSReply(s32 return_value, u64 extra_tb_ticks = 0)
+{
+  // According to hardware tests, FS takes at least 2700 TB ticks to reply to commands.
+  return {return_value, true, (2700 + extra_tb_ticks) * SystemTimers::TIMER_RATIO};
+}
+
+/// Amount of TB ticks required for a superblock write to complete.
+constexpr u64 SUPERBLOCK_WRITE_TICKS = 3370000;
+/// Amount of TB ticks required to write a cluster (for a file).
+constexpr u64 CLUSTER_WRITE_TICKS = 300000;
+/// Amount of TB ticks required to read a cluster (for a file).
+constexpr u64 CLUSTER_READ_TICKS = 115000;
+constexpr size_t CLUSTER_DATA_SIZE = 0x4000;
+
+FS::FS(Kernel& ios, const std::string& device_name) : Device(ios, device_name)
+{
+}
+
+void FS::DoState(PointerWrap& p)
+{
+  p.Do(m_fd_map);
+  p.Do(m_cache_fd);
+  p.Do(m_cache_chain_index);
+  p.Do(m_dirty_cache);
+}
+
+static void LogResult(const std::string& command, ResultCode code)
+{
+  GENERIC_LOG(LogTypes::IOS_FS, (code == ResultCode::Success ? LogTypes::LINFO : LogTypes::LERROR),
+              "%s: result %d", command.c_str(), ConvertResult(code));
+}
+
+template <typename T>
+static void LogResult(const std::string& command, const Result<T>& result)
+{
+  LogResult(command, result.Succeeded() ? ResultCode::Success : result.Error());
+}
+
+enum class FileLookupMode
+{
+  Normal,
+  /// Timing model to use when FS splits the path into a parent and a file name
+  /// and looks up each of them individually.
+  Split,
+};
+static u64 EstimateFileLookupTicks(const std::string& path, FileLookupMode mode)
+{
+  const size_t number_of_path_components = std::count(path.cbegin(), path.cend(), '/');
+  if (number_of_path_components == 0)
+    return 0;
+  // Paths that end with a slash are invalid and rejected early in FS.
+  if (!path.empty() && *path.rbegin() == '/')
+    return 300;
+  if (mode == FileLookupMode::Normal)
+    return 680 * number_of_path_components;
+  return 1000 + 340 * number_of_path_components;
+}
+
+IPCCommandResult FS::Open(const OpenRequest& request)
+{
+  if (m_fd_map.size() >= 16)
+    return GetFSReply(ConvertResult(ResultCode::NoFreeHandle));
+
+  if (request.path.size() >= 64)
+    return GetFSReply(ConvertResult(ResultCode::Invalid));
+
+  if (request.path == "/dev/fs")
+  {
+    m_fd_map[request.fd] = {request.gid, request.uid, INVALID_FD};
+    return GetFSReply(IPC_SUCCESS);
+  }
+
+  const u64 ticks = EstimateFileLookupTicks(request.path, FileLookupMode::Normal);
```

Comment:
Lookup will stop at the first non existing path (as IOS won't be able to traverse the FS any further).

Currently we only simulate the lookup timing approximately, because making that more accurate would require checking every path component for existence. I'll add a comment for this.